pkg/cloudprovider/providers/gce/gce_instances.go

```
@@ -274,12 +275,55 @@ func (gce *GCECloud) AddSSHKeyToAllInstances(user string, keyData []byte) error
 	})
 }
 
-// GetAllCurrentZones returns all the zones in which nodes are currently running
-// TODO: Caching, but this is currently only called when we are creating a volume,
-// which is a relatively infrequent operation, and this is only 1 API call.
-// Ideally we want to make this watch-based
+// GetAllCurrentZones returns all the zones in which k8s nodes are currently running
 func (gce *GCECloud) GetAllCurrentZones() (sets.String, error) {
-	return gce.manager.GetAllCurrentZones(gce.client)
+	if gce.nodeInformerSynced == nil {
+		glog.Warningf("GCECloud object does not have informers set, should only happen in E2E tests.")
+		return gce.GetAllZonesE2E()
+	}
+	if !controller.WaitForCacheSync("cloud-provider", wait.NeverStop, gce.nodeInformerSynced) {
```

I don't know if I would wait forever.  Can we just return an error immediately?